server_cached_router.go

```diff
@@ -0,0 +1,230 @@
+// It will consult cache ONLY if the addrs slice passed to it is empty.
+func (r cachedRouter) withAddrsFromCache(queryOrigin string, pid peer.ID, addrs []types.Multiaddr) []types.Multiaddr {
+	// skip cache if we already have addrs
+	if len(addrs) > 0 {
+		peerAddrLookups.WithLabelValues(addrCacheStateUnused, queryOrigin).Inc()
+		return addrs
+	}
+
+	cachedAddrs := r.cachedAddrBook.GetCachedAddrs(pid)
+	if len(cachedAddrs) > 0 {
+		logger.Debugw("found cached addresses", "peer", pid, "cachedAddrs", cachedAddrs)
+		peerAddrLookups.WithLabelValues(addrCacheStateHit, queryOrigin).Inc()
+		return cachedAddrs
+	} else {
+		// Cache miss. Queue peer for lookup.
+		peerAddrLookups.WithLabelValues(addrCacheStateMiss, queryOrigin).Inc()
+		return nil
+	}
+}
+
+var _ iter.ResultIter[types.Record] = &cacheFallbackIter{}
+
+type cacheFallbackIter struct {
+	sourceIter      iter.ResultIter[types.Record]
+	current         iter.Result[types.Record]
+	findPeersResult chan types.PeerRecord
+	router          cachedRouter
+	ctx             context.Context
+	cancel          context.CancelFunc
+	ongoingLookups  atomic.Int32
+}
+
+// NewCacheFallbackIter is a wrapper around a results iterator that will resolve peers with no addresses from cache and if no cached addresses, will look them up via FindPeers.
+// It's a bit complex because it ensures we continue iterating without blocking on the FindPeers call.
+func NewCacheFallbackIter(sourceIter iter.ResultIter[types.Record], router cachedRouter, ctx context.Context) *cacheFallbackIter {
+	ctx, cancel := context.WithCancel(ctx)
+	return &cacheFallbackIter{
+		sourceIter:      sourceIter,
+		router:          router,
+		ctx:             ctx,
+		cancel:          cancel,
+		findPeersResult: make(chan types.PeerRecord),
+		ongoingLookups:  atomic.Int32{},
+	}
+}
+
+func (it *cacheFallbackIter) Next() bool {
+	// Try to get the next value from the source iterator first
+	if it.sourceIter.Next() {
+		val := it.sourceIter.Val()
+		handleRecord := func(id *peer.ID, record *types.PeerRecord) bool {
+			record.Addrs = it.router.withAddrsFromCache(addrQueryOriginProviders, *id, record.Addrs)
+			if record.Addrs != nil {
+				it.current = iter.Result[types.Record]{Val: record}
+				return true
+			}
+			logger.Infow("no cached addresses found in cacheFallbackIter, dispatching find peers", "peer", id)
+
+			if it.router.cachedAddrBook.ShouldProbePeer(*id) {
+				it.ongoingLookups.Add(1) // important to increment before dispatchFindPeer
+				// If a record has no addrs, we dispatch a lookup to find addresses
+				go it.dispatchFindPeer(*record)
+
+				return it.Next() // Recursively call Next() to either read from sourceIter or wait for lookup result
+			}
+
+			// If we're not going to probe, return the record with no addrs
+			// TODO: should we even return these if the peer is likely unreachable?
+			it.current = iter.Result[types.Record]{Val: record}
+			return true
```

Comment:
:warning: if we did active probing before, we should skip those empty peers for sure:

```suggestion
			}
			return it.Next() // Recursively call Next() to either read from sourceIter or wait for lookup result
```


Rationale:

(in most cases they will cause clients to do useless peer lookups that will fail as well)

Currently (with those empty results) we have bizzare behavior where initial  lookup returns only online peers, but subsequent lookup starts returning offline peers that failed initial probe.

Practical example: lets find providers for /ipns/protocol.ai:

```console
$ curl "http://127.0.0.1:8190/routing/v1/providers/QmbKiad69C9Ucy3enEyKpmeyiKk2NPi628dDs3PanTKR1P" -H 'Accept: application/x-ndjson' -s  
{"Addrs":["/ip4/69.203.71.93/udp/64283/quic-v1/p2p/12D3KooWNMURc4AuvjC1vYARFTSKmMJBB1pjyr5zymBL15E96rqA/p2p-circuit","/ip4/95.216.28.167/udp/4001/quic/p2p/12D3KooWPDBio8GH9svF7oS1mw849esxLs3gr5YViUGqzi2zwef9/p2p-circuit","/ip4/95.216.28.167/tcp/4001/p2p/12D3KooWPDBio8GH9svF7oS1mw849esxLs3gr5YViUGqzi2zwef9/p2p-circuit","/ip4/69.203.71.93/udp/64283/quic-v1/webtransport/certhash/uEiAZ3F5UJwcCIr3UPr_sPuWQj3v4HAasik_H_dLt0bzZTQ/certhash/uEiAPUcxZdKwDMsPCOs8qe4rmvLJRdNUVH6Mhi2nSy6EniQ/p2p/12D3KooWNMURc4AuvjC1vYARFTSKmMJBB1pjyr5zymBL15E96rqA/p2p-circuit","/ip4/69.203.71.93/tcp/64283/p2p/12D3KooWNMURc4AuvjC1vYARFTSKmMJBB1pjyr5zymBL15E96rqA/p2p-circuit"],"ID":"12D3KooWSXS5KPLpPZ4jfCfek5SrCW2EeGFfkQr3fqb3UAikfX58","Schema":"peer"}
{"Addrs":["/ip4/216.238.81.231/udp/4001/quic-v1/p2p/12D3KooWRo1uYUSm9wXn2LeBpZyQrU3KYeq2yzkKRSGc8FYjDXpb/p2p-circuit","/ip4/216.238.81.231/tcp/4001/p2p/12D3KooWRo1uYUSm9wXn2LeBpZyQrU3KYeq2yzkKRSGc8FYjDXpb/p2p-circuit","/ip4/216.238.81.231/udp/4001/quic/p2p/12D3KooWRo1uYUSm9wXn2LeBpZyQrU3KYeq2yzkKRSGc8FYjDXpb/p2p-circuit","/ip4/66.29.147.177/tcp/4001/p2p/12D3KooWJNX7YDRBkh1m2vssumMHoP7PEn2gQ7WVs5s3vYxGfxxb/p2p-circuit","/ip4/66.29.147.177/udp/4001/quic-v1/webtransport/certhash/uEiBjZkayyaLxp88App6om86GYU2q-TB8I50gowq5MCCPqQ/certhash/uEiCwAXs3XknzDnLuXHkvwiJn-StgfExwnGPR-7pA5zsrmg/p2p/12D3KooWJNX7YDRBkh1m2vssumMHoP7PEn2gQ7WVs5s3vYxGfxxb/p2p-circuit","/ip4/66.29.147.177/udp/4001/quic-v1/p2p/12D3KooWJNX7YDRBkh1m2vssumMHoP7PEn2gQ7WVs5s3vYxGfxxb/p2p-circuit"],"ID":"12D3KooWEp8H3jDYw1Cn3hL432UX71Wv4UKmTXHP4mmL597GTmVn","Schema":"peer"}
{"Addrs":["/ip4/145.40.89.101/tcp/4001","/ip6/2604:1380:45f1:d800::1/tcp/4001"],"ID":"12D3KooWHVXoJnv2ifmr9K6LWwJPXxkfvzZRHzjiTZMvybeTnwPy","Schema":"peer"}
{"Addrs":["/ip4/145.40.89.101/udp/4001/quic-v1/webtransport/certhash/uEiBO8uH1QnET8rmwLLXlR5rocnu8gjZ1TZDQdMdp_yOowA/certhash/uEiA7PuBdt9syTOi4s3zSDE1XD2kBGN__66nSYaszeH3xRA","/ip6/2604:1380:45f1:d800::1/tcp/4002/tls/sni/2604-1380-45f1-d800--1.k51qzi5uqu5dj0wvrbb8keygfyxe2v0fi1qbqz4pl3zzozle7oaqhf97mqazo4.libp2p.direct/ws","/ip4/145.40.89.101/tcp/4002/tls/sni/145-40-89-101.k51qzi5uqu5dj0wvrbb8keygfyxe2v0fi1qbqz4pl3zzozle7oaqhf97mqazo4.libp2p.direct/ws","/ip4/145.40.89.101/udp/4001/quic-v1","/ip6/2604:1380:45f1:d800::1/tcp/4001","/ip6/2604:1380:45f1:d800::1/udp/4001/webrtc-direct/certhash/uEiAE_vuhjpPz5yq87PvAUAWca2yJxiuByzd_wRKiqcJGlw","/ip6/2604:1380:45f1:d800::1/udp/4001/quic-v1/webtransport/certhash/uEiBO8uH1QnET8rmwLLXlR5rocnu8gjZ1TZDQdMdp_yOowA/certhash/uEiA7PuBdt9syTOi4s3zSDE1XD2kBGN__66nSYaszeH3xRA","/ip4/145.40.89.101/udp/4001/webrtc-direct/certhash/uEiAE_vuhjpPz5yq87PvAUAWca2yJxiuByzd_wRKiqcJGlw","/ip4/145.40.89.101/tcp/4001","/ip6/2604:1380:45f1:d800::1/udp/4001/quic-v1"],"ID":"12D3KooWHVXoJnv2ifmr9K6LWwJPXxkfvzZRHzjiTZMvybeTnwPy","Schema":"peer"}
```

Then calling second time returns empty ones, likely  random people who visited /ipns/protocol.ai with their desktop node that is currently offline:

```console
$ curl "http://127.0.0.1:8190/routing/v1/providers/QmbKiad69C9Ucy3enEyKpmeyiKk2NPi628dDs3PanTKR1P" -H 'Accept: application/x-ndjson' -s
{"ID":"12D3KooWAMhEcutSZRHP9BiSKa8Tsz3cFefKBSezhJRqyb62rLFg","Schema":"peer"}
{"ID":"12D3KooWCUW1rNfUtyJCATxCh1yQhVdwMUJYywLYWk5X9FgTxrPT","Schema":"peer"}
{"ID":"12D3KooWCoqS7A7MBy7pbAhJbPNwXytwYkBC12A7h3frUe3Tip6p","Schema":"peer"}
{"ID":"12D3KooWGqzJNT8RSg5YPpProaeYqXQKLg8k7grChnyL4dMp1sED","Schema":"peer"}
{"ID":"12D3KooWJiVts67y1rbQbKJMuKLzLBjnMu7xLB7o6BxXMBNp1pNG","Schema":"peer"}
{"ID":"12D3KooWPJaWvf5W8Zhss97CvDzs9LpNuewHFvUVZTFFkyAcHcM2","Schema":"peer"}
{"ID":"12D3KooWQCbo62Cs8uXPJPdrW3S7DQCbSABezR3mMud4uSzetMeT","Schema":"peer"}
{"ID":"12D3KooWQa3hz3LsGMVfsUSwJUFurekidxAw8QaNyrqxXHr3iE2m","Schema":"peer"}
{"ID":"12D3KooWDot6SE3cVqAPe6Rwk1zunSrUTg9Z1RmB5YHgouEnQ6Zx","Schema":"peer"}
{"ID":"12D3KooWE3CpyihkLaSGAZEyEJ9sWijByTZ4HHqiC4w3thqufoSs","Schema":"peer"}
{"ID":"12D3KooW9x8nmeaP2i4SZD27Q6Ew6zNrffBnbLnMqCmzzW9PtZtF","Schema":"peer"}
{"ID":"12D3KooW9uKfmrqDXmiKAcGw3Q5csKjMPBcg7aBgXPToe8GpYRun","Schema":"peer"}
{"ID":"12D3KooW9yycnejq1rRXc16bUGBQWDeLPXd81UG4LwRPg621uYVX","Schema":"peer"}
{"ID":"12D3KooWA2viJLxkVNrs8v1FW59q8qopR99GueJ11YYzSpT9wVZo","Schema":"peer"}
{"ID":"12D3KooWARy1rzqfHU1w3aHgXxpk1WpmAcRUVqc31xcwr5ZmYhu3","Schema":"peer"}
{"ID":"12D3KooWAX7yomznUC1HXAh2MuXkAxCSGkBA3FsUHhS53PvSc7A7","Schema":"peer"}
{"ID":"12D3KooWCPgSBDH637g5sAxJRLuSngefzMGUeNJEtJkjabk3qmB1","Schema":"peer"}
{"ID":"12D3KooWCq9R7DmRbxgwHYc6FwzmJL6tZMK1qMhTVqWbmf8GbSvZ","Schema":"peer"}
{"ID":"12D3KooWEc9p8exRiDzeNofnxASWhk7r9BEUgY8rPYdHJ8MGmE8u","Schema":"peer"}
{"ID":"12D3KooWFnbMPGB2u26Z8Kv2ZTxoHZaucSUczJEH9xXogjnnSxsV","Schema":"peer"}
{"ID":"12D3KooWFogZTo5LTsveft62VRGjVCFkhPGWrSyvdPy5xkuVrWPZ","Schema":"peer"}
{"ID":"12D3KooWFqq1yzb65kWTp9oMXyYS9HuVmCYxj8STCf9TiVpJXogr","Schema":"peer"}
{"ID":"12D3KooWGjcvV93ThR9E1HgsEDG7yzbKbjHBj6iZusDpkg9yNR44","Schema":"peer"}
{"ID":"12D3KooWH5xqQ39okv5PpxwdPWCNyWgLqJ6vv58eyo5auqDNJTy2","Schema":"peer"}
{"ID":"12D3KooWNDXvi4vQn8xBtQXCcMiUA65zY9yfJF7FQehjtKA2Yv4d","Schema":"peer"}
{"Addrs":["/ip4/127.0.0.1/tcp/4001","/ip6/2604:1380:45f1:d800::1/udp/4001/quic-v1/webtransport/certhash/uEiBO8uH1QnET8rmwLLXlR5rocnu8gjZ1TZDQdMdp_yOowA/certhash/uEiA7PuBdt9syTOi4s3zSDE1XD2kBGN__66nSYaszeH3xRA","/ip6/::1/udp/4001/webrtc-direct/certhash/uEiAE_vuhjpPz5yq87PvAUAWca2yJxiuByzd_wRKiqcJGlw","/ip4/145.40.89.101/udp/4001/webrtc-direct/certhash/uEiAE_vuhjpPz5yq87PvAUAWca2yJxiuByzd_wRKiqcJGlw","/ip4/145.40.89.101/udp/4001/quic-v1","/ip4/127.0.0.1/udp/4001/quic-v1/webtransport/certhash/uEiBO8uH1QnET8rmwLLXlR5rocnu8gjZ1TZDQdMdp_yOowA/certhash/uEiA7PuBdt9syTOi4s3zSDE1XD2kBGN__66nSYaszeH3xRA","/ip4/145.40.89.101/tcp/4001","/ip4/145.40.89.101/tcp/4002/tls/sni/145-40-89-101.k51qzi5uqu5dj0wvrbb8keygfyxe2v0fi1qbqz4pl3zzozle7oaqhf97mqazo4.libp2p.direct/ws","/ip6/::1/udp/4001/quic-v1","/ip6/::1/udp/4001/quic-v1/webtransport/certhash/uEiBO8uH1QnET8rmwLLXlR5rocnu8gjZ1TZDQdMdp_yOowA/certhash/uEiA7PuBdt9syTOi4s3zSDE1XD2kBGN__66nSYaszeH3xRA","/ip6/2604:1380:45f1:d800::1/udp/4001/webrtc-direct/certhash/uEiAE_vuhjpPz5yq87PvAUAWca2yJxiuByzd_wRKiqcJGlw","/ip4/127.0.0.1/udp/4001/webrtc-direct/certhash/uEiAE_vuhjpPz5yq87PvAUAWca2yJxiuByzd_wRKiqcJGlw","/ip4/127.0.0.1/udp/4001/quic-v1","/ip6/2604:1380:45f1:d800::1/udp/4001/quic-v1","/ip6/2604:1380:45f1:d800::1/tcp/4001","/ip6/2604:1380:45f1:d800::1/tcp/4002/tls/sni/2604-1380-45f1-d800--1.k51qzi5uqu5dj0wvrbb8keygfyxe2v0fi1qbqz4pl3zzozle7oaqhf97mqazo4.libp2p.direct/ws","/ip4/145.40.89.101/udp/4001/quic-v1/webtransport/certhash/uEiBO8uH1QnET8rmwLLXlR5rocnu8gjZ1TZDQdMdp_yOowA/certhash/uEiA7PuBdt9syTOi4s3zSDE1XD2kBGN__66nSYaszeH3xRA","/ip6/::1/tcp/4001"],"ID":"12D3KooWHVXoJnv2ifmr9K6LWwJPXxkfvzZRHzjiTZMvybeTnwPy","Schema":"peer"}
{"ID":"12D3KooWJwuFqk7MkGDBbQ3Pz9EKQ3qn7m93SqZQFMNSq3bbp4Tq","Schema":"peer"}
{"ID":"12D3KooWLW7F5Kdrt21kPGKezLbJ2UTPuet7V6oQVJW1Z3VU1wdi","Schema":"peer"}
{"ID":"12D3KooWMbc9bbLBTFSmke99ZJETWTiSg3arCbMG6uZ7LMkX7WiG","Schema":"peer"}
{"ID":"12D3KooWMo7ZLNLxaCpGFHVJLEVDq6grrjULqvZ9eGuqY6DWuCAP","Schema":"peer"}
{"ID":"12D3KooWMqEqQiL5bsAE56Xg5nkWBW8o6cT6q7yvkjvCA5fnsAU2","Schema":"peer"}
{"ID":"12D3KooWPShtsFRTLtCJLpNVxjysSpmmzyy18rP4k7rh2TaDre5J","Schema":"peer"}
{"ID":"12D3KooWPajM7tGMpwvaDrBtVLsDmvoUMuqySm4tRTayowe1RJsQ","Schema":"peer"}
{"ID":"12D3KooWQCsg7PXQNQUGs2hJW7m1kXTr1Lt6MRgqwR8YazTcwEPP","Schema":"peer"}
{"ID":"12D3KooWQKKQCtxgQ3BNoTj5TgFxZywo7SimYK3Zyjr1K9EUnYes","Schema":"peer"}
{"Addrs":["/ip4/95.216.28.167/udp/4001/quic/p2p/12D3KooWPDBio8GH9svF7oS1mw849esxLs3gr5YViUGqzi2zwef9/p2p-circuit","/ip4/95.216.28.167/tcp/4001/p2p/12D3KooWPDBio8GH9svF7oS1mw849esxLs3gr5YViUGqzi2zwef9/p2p-circuit","/ip4/69.203.71.93/udp/64283/quic-v1/webtransport/certhash/uEiAZ3F5UJwcCIr3UPr_sPuWQj3v4HAasik_H_dLt0bzZTQ/certhash/uEiAPUcxZdKwDMsPCOs8qe4rmvLJRdNUVH6Mhi2nSy6EniQ/p2p/12D3KooWNMURc4AuvjC1vYARFTSKmMJBB1pjyr5zymBL15E96rqA/p2p-circuit","/ip4/69.203.71.93/tcp/64283/p2p/12D3KooWNMURc4AuvjC1vYARFTSKmMJBB1pjyr5zymBL15E96rqA/p2p-circuit","/ip4/69.203.71.93/udp/64283/quic-v1/p2p/12D3KooWNMURc4AuvjC1vYARFTSKmMJBB1pjyr5zymBL15E96rqA/p2p-circuit"],"ID":"12D3KooWSXS5KPLpPZ4jfCfek5SrCW2EeGFfkQr3fqb3UAikfX58","Schema":"peer"}
{"Addrs":["/ip4/66.29.147.177/tcp/4001/p2p/12D3KooWJNX7YDRBkh1m2vssumMHoP7PEn2gQ7WVs5s3vYxGfxxb/p2p-circuit","/ip4/66.29.147.177/udp/4001/quic-v1/webtransport/certhash/uEiBjZkayyaLxp88App6om86GYU2q-TB8I50gowq5MCCPqQ/certhash/uEiCwAXs3XknzDnLuXHkvwiJn-StgfExwnGPR-7pA5zsrmg/p2p/12D3KooWJNX7YDRBkh1m2vssumMHoP7PEn2gQ7WVs5s3vYxGfxxb/p2p-circuit","/ip4/66.29.147.177/udp/4001/quic-v1/p2p/12D3KooWJNX7YDRBkh1m2vssumMHoP7PEn2gQ7WVs5s3vYxGfxxb/p2p-circuit","/ip4/216.238.81.231/udp/4001/quic-v1/p2p/12D3KooWRo1uYUSm9wXn2LeBpZyQrU3KYeq2yzkKRSGc8FYjDXpb/p2p-circuit","/ip4/216.238.81.231/tcp/4001/p2p/12D3KooWRo1uYUSm9wXn2LeBpZyQrU3KYeq2yzkKRSGc8FYjDXpb/p2p-circuit","/ip4/216.238.81.231/udp/4001/quic/p2p/12D3KooWRo1uYUSm9wXn2LeBpZyQrU3KYeq2yzkKRSGc8FYjDXpb/p2p-circuit"],"ID":"12D3KooWEp8H3jDYw1Cn3hL432UX71Wv4UKmTXHP4mmL597GTmVn","Schema":"peer"}
```

Reply:
I agree we should not return providers without addresses if we probed them.

At first this behaviour you caught here seemed was strange, but then I realised that this happens because we pass the context from the http request all the way to the additional `FindPeers`, which results in those `FindPeers` not completing due to an exceeded context deadline after the [default 30 seconds](https://github.com/ipfs/boxo/commit/e38f236348d62c9c75494c563975e3eccc745b10) which we recently added by default. 

A couple of options:
- Decouple the dispatched FindPeer from the request context. This is tricky to do, because the router here is an iterator rather than a channel, so it's a bit hard to handle this concurrently. 
- Change the default timeout
-